HDDS-8659. Ozone admin SCM CLI command for block tokens

tanvipenumudy opened the pull request:
## What changes were proposed in this pull request?

Introduce an Ozone admin SCM CLI command to force generate new secret keys (rotate).

Format: `ozone admin scm rotate --scm hostPort`

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8659

## How was this patch tested?

Added integration test cases.

tanvipenumudy commented:
@duongkame, @kerneltime, @sadanand48, @siddhantsangwan could you please take a look? Thanks

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/RotateKeySubCommand.java:
```diff
@@ -0,0 +1,40 @@
+
+  @Override
+  protected void execute(ScmClient scmClient) throws IOException {
+    try (ScmClient client = new ContainerOperationClient(
```
We can use the same scmClient here from the method , for the unit test to work , we could just pass a list  of config args in the args array in --set=key=value format.

tanvipenumudy replied:
Thanks @sadanand48, made the required changes!

sadanand48 commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerOperationClient.java:
```diff
@@ -124,6 +129,16 @@ public static StorageContainerLocationProtocol newContainerRpcClient(
+    try {
+      return HddsServerUtil.getScmSecretClient(configSource);
+    } catch (IOException e) {
+      e.printStackTrace();
```
LOG.error instead

sadanand48 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:
```diff
@@ -232,6 +234,16 @@ private void validateSecretKeyStatus() throws SCMSecretKeyException {
+    try {
+      validateSecretKeyStatus();
+    } catch (SCMSecretKeyException e) {
+      e.printStackTrace();
```
LOG.error instead

duongkame replied:
This error should be thrown to client-side, like other APIs.

sadanand48 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/SecretKeyProtocolServerSideTranslatorPB.java:
```diff
@@ -102,6 +108,16 @@ public SCMSecretKeyResponse processRequest(SCMSecretKeyRequest request)
+                  checkAndRotate(request.getCheckAndRotateRequest().getForce()))
+              .build();
+        } catch (TimeoutException e) {
+          e.printStackTrace();
```
LOG.error instead

duongkame replied:
+1 error should be handled by the client (CLI)

sadanand48 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestBlockTokens.java:
```diff
@@ -377,6 +386,24 @@ private static void setSecureConfig() throws IOException {
+
+    String oldKey =
+        scmClient.getSecretKeyClient().getCurrentSecretKey().toString();
+    Thread.sleep(1000);
```
Since the force flag rotates the key irrespective of whether duration has expired, this sleep might not be needed.

duongkame commented:
> Format: ozone admin scm rotate --scm hostPort

I think the CLI should not rely on the admin to find the destination SCM, instead itself find the leader using the failover proxy, and only the leader SCM should accept this call, and process this call.

Also, the CLI can be divided into 2 usages.

```
ozone admin scm rotate
# this command only trigger a rotation if the the latest secret key timestamp already pass the rotate duration
# otherwise, it prints out a message to explain why rotation doesn't happen, e.g. `The lastest key was created at {lastest key timestamp} which doesn't pass the rotate duration of 1 day yet`.
```

```
ozone admin scm rotate --force
# force key rotation to happen now.
```

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocolPB/SecretKeyProtocolClientSideTranslatorPB.java:
```diff
@@ -129,6 +131,23 @@ public ManagedSecretKey getCurrentSecretKey() throws IOException {
+          builder.setCheckAndRotateRequest(request))
+          .getCheckAndRotateResponseProto().getStatus();
+    } catch (IOException e) {
+      e.printStackTrace();
```
should be thrown and handled outside.

tanvipenumudy commented:
> I think the CLI should not rely on the admin to find the destination SCM, instead itself find the leader using the failover proxy, and only the leader SCM should accept this call, and process this call.

The command should work even by skipping the `--scm` flag.

> Also, the CLI can be divided into 2 usages.

Sure, will be making the changes for adding an explicit `--force` flag to the SCM CLI.

tanvipenumudy commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocolPB/SecretKeyProtocolClientSideTranslatorPB.java:
```diff
@@ -50,7 +52,7 @@
  */
 public class SecretKeyProtocolClientSideTranslatorPB implements
-    SecretKeyProtocol, ProtocolTranslator, Closeable {
+    SecretKeyProtocol, SecretKeyProtocolScm, ProtocolTranslator, Closeable {
```
Thank you, I have made the changes to avoid redundant implementation.

duongkame commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/client/ScmClient.java:
```diff
@@ -362,6 +363,15 @@ StartContainerBalancerResponseProto startContainerBalancer(
+   * @return
+   * @throws TimeoutException
+   */
+  boolean checkAndRotate(boolean force) throws TimeoutException, IOException;
```
nit: maybe name it `rotateSecretKeys`. Note that this class name has no SecretKey context, so the method name needs it.

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/HddsServerUtil.java:
```diff
@@ -511,6 +512,18 @@ public static SCMSecurityProtocol getScmSecurityClient(
         SCMSecurityProtocol.class, conf);
   }
 
+  public static SecretKeyProtocolScm getScmSecretClient(
```
Add a javadoc to mention it's intended to be used by clients under SCM identify. Maybe rename it to `getSecretKeyClientForSCM` to make the intention clear.

duongkame commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/SecretKeyProtocolServerSideTranslatorPB.java:
```diff
@@ -102,6 +104,16 @@ public SCMSecretKeyResponse processRequest(SCMSecretKeyRequest request)
+                  checkAndRotate(request.getCheckAndRotateRequest().getForce()))
+              .build();
+        } catch (TimeoutException e) {
+          LOG.error("Timeout occurred while executing checkAndRotate.", e);
```
See the exception handling several lines down. I think this timeout error can be considered an internal error and get transferred to client side.

duongkame commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:
```diff
@@ -232,6 +233,16 @@ private void validateSecretKeyStatus() throws SCMSecretKeyException {
+    try {
+      validateSecretKeyStatus();
+    } catch (SCMSecretKeyException e) {
+      LOGGER.error("Error while validating the secret key status.", e);
```
Don't catch these validation exceptions on server side. We should let the translator to translate them to error code and transfer to client side.

duongkame commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerOperationClient.java:
```diff
@@ -124,6 +129,16 @@ public static StorageContainerLocationProtocol newContainerRpcClient(
+    try {
+      return HddsServerUtil.getScmSecretClient(configSource);
+    } catch (IOException e) {
+      LOG.error("Error while getting the SCM secret client.", e);
```
Just throw, let the client code (the CLI catch and log it).

duongkame commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestBlockTokens.java:
```diff
@@ -377,6 +386,23 @@ private static void setSecureConfig() throws IOException {
   }
 
+  @Test
+  public void testRotateKeySCMAdminCommand()
```
I think we need to move this test to a new test class. `TestBlockToken` is intended for block token test-cases and configure a relatively short secret key lifetime. This CLI test should be done with a longer rotation duration, e.g. 1h, so that the CLI and the automatic rotation don't collide.

ashishkumar50 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/RotateKeySubCommand.java:
```diff
@@ -0,0 +1,57 @@
+      }
+      System.out.println(
+          "Secret key rotation is complete, a new key has been generated. " +
+              "Rotate Status: " + status);
```
When there is TimeoutException, This message will not be correct to show. Instead we can print different message in catch and return from there.

tanvipenumudy replied:
Thank you for pointing this, made the required changes!

ashishkumar50 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestBlockTokens.java:
```diff
@@ -377,6 +386,23 @@ private static void setSecureConfig() throws IOException {
+    String[] args = {"scm", "rotate", "--scm", hostPort};
+
+    String oldKey =
+        scmClient.getSecretKeyClient().getCurrentSecretKey().toString();
```
May be we can add a small test here, run twice "getCurrentSecretKey" and assert whether both are same before rotating.

ashishkumar50 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/RotateKeySubCommand.java:
```diff
@@ -0,0 +1,57 @@
+      boolean status = false;
+      try {
+        status = client.checkAndRotate(true);
+      } catch (TimeoutException e) {
```
checkAndRotate can even throw IOException, we can handle and show proper message too.

ashishkumar50 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/RotateKeySubCommand.java:
```diff
@@ -0,0 +1,57 @@
+              "Rotate Status: " + status);
+    }
+  }
+}
```
Add a new line at the end of file.

ashishkumar50 commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocol/SecretKeyProtocolScm.java:
```diff
@@ -28,4 +31,13 @@
+   *
+   * @param force boolean flag that forcefully rotates the key on demand
+   * @return
+   * @throws TimeoutException
```
Please Correct javadoc for @return and @throws.

ashishkumar50 commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/SecretKeyManager.java:
```diff
@@ -109,12 +109,13 @@ public boolean isInitialized() {
 
     ManagedSecretKey currentKey = state.getCurrentKey();
-    if (shouldRotate(currentKey)) {
+    if (shouldRotate(currentKey) || force) {
```
nit: we can reverse condition "if (force || shouldRotate(currentKey))", to avoid method call if force is true.

ashishkumar50 commented on hadoop-hdds/interface-server/src/main/proto/ScmSecretKeyProtocol.proto:
```diff
@@ -98,6 +103,11 @@ message SCMGetSecretKeyRequest {
 }
 
+message SCMGetCheckAndRotateRequest {
+    optional bool force = 1;
```
We can change to optional bool force = 1 [default = false];

ashishkumar50 reviewed (commented):
Hi @tanvipenumudy, Thanks for working on this. Please find few comments.

adoroszlai commented:
@tanvipenumudy CI runs in forks for each push.  There is some limit on the concurrent Github Actions jobs/workflows for each repo (although I don't know the details).  If you push in multiple batches, you might want to cancel [runs](https://github.com/tanvipenumudy/ozone/actions/workflows/post-commit.yml?query=branch%3AHDDS-8659) for outdated versions of your code, otherwise the latest commits will be queued for a long time.

tanvipenumudy commented:
Thank you @adoroszlai, noted. Will be cancelling all the previous runs to enable quicker Git CI/CD checks over the latest commits.

tanvipenumudy commented:
Thank you for the reviews, I have addressed the comments on the PR. 

The [Git CI: integration (ozone)](https://github.com/tanvipenumudy/ozone/actions/workflows/post-commit.yml?query=branch%3AHDDS-8659) seems to fail at `org.apache.hadoop.ozone.TestSecretKeysApi` - but the test cases have been passing locally.

duongkame commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/SecretKeyProtocolServerSideTranslatorPB.java:
```diff
@@ -102,6 +104,21 @@ public SCMSecretKeyResponse processRequest(SCMSecretKeyRequest request)
+              .setCheckAndRotateResponseProto(
+                  checkAndRotate(request.getCheckAndRotateRequest().getForce()))
+              .build();
+        } catch (TimeoutException e) {
```
This exception handling can happen in `SCMSecurityProtocolServer.checkAndRotate`, so that we can avoid TimeoutException in the interface level.

duongkame commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:
```diff
@@ -232,6 +233,13 @@ private void validateSecretKeyStatus() throws SCMSecretKeyException {
+  public boolean checkAndRotate(boolean force)
+      throws TimeoutException, SCMSecretKeyException {
+    validateSecretKeyStatus();
+    return secretKeyManager.checkAndRotate(force);
```
```suggestion
    try {
      return secretKeyManager.checkAndRotate(force);
    } catch (TimeoutException ex) {
      LOGGER.error("Timeout rotating secret keys", ex);
      throw new SCMSecretKeyException(ex.getMessage(), SCMSecretKeyException.ErrorCode.INTERNAL_ERROR);
    }
```

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocol/SecretKeyProtocolScm.java:
```diff
@@ -28,4 +31,14 @@
+   * @throws TimeoutException
+   * @throws IOException
+   */
+  boolean checkAndRotate(boolean force) throws TimeoutException, IOException;
```
As below comments, `TimeoutException` should not be exposed to client side.

tanvipenumudy replied:
Understood, thanks @duongkame, made the required changes!

duongkame reviewed (commented):
Thanks for the patch @tanvipenumudy 
LGTM, just left one comment to avoid exposing TimeoutException in client apis.

tanvipenumudy commented:
Thank you @duongkame, made the required changes to avoid exposing `TimeoutException`.

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/SecretKeyManager.java:
```diff
@@ -126,6 +127,8 @@ public synchronized boolean checkAndRotate() throws SCMException {
       state.updateKeys(updatedKeys);
       return true;
     }
+    LOG.info("The latest key was created at: " + currentKey.getCreationTime() +
```
We may not need this log as INFO, as it'll blood the log files with not-very-useful messages. Note that this function is called by a periodic schedule.

tanvipenumudy replied:
Understood, will remove this log message, thanks.

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/SecretKeyManager.java:
```diff
@@ -109,12 +109,13 @@ public boolean isInitialized() {
 
     ManagedSecretKey currentKey = state.getCurrentKey();
-    if (shouldRotate(currentKey)) {
+    if (force || shouldRotate(currentKey)) {
```
If the rotate is a result of a force, it'll be useful to add a log line to state that.

tanvipenumudy replied:
Thanks @duongkame, I have added the log message for the same.

duongkame reviewed (approved):
Thanks @tanvipenumudy for the patch. +1. 
I just leave some minor comments related to logging. Please address them before merging.

tanvipenumudy commented on hadoop-hdds/interface-server/src/main/proto/ScmSecretKeyProtocol.proto:
```diff
@@ -67,12 +69,15 @@ message SCMSecretKeyResponse {
     GetCurrentSecretKey = 1;
     GetSecretKey = 2;
     GetAllSecretKeys = 3;
+    GetCheckAndRotate = 4;
```
Thanks, I have renamed the field!

sadanand48 reviewed (commented):
Thanks @tanvipenumudy for the patch , @duongkame, @ashishkumar50 , @krishnaasawa1 for the reviews.
